warehouse/query-core/src/main/java/datawave/query/util/sortedmap/MultiMapBackedSortedMap.java

```diff
@@ -0,0 +1,397 @@
+        }
+
+        @Override
+        public void remove() {
+            if (!populated) {
+                throw new IllegalStateException();
+            }
+            Exception e = null;
+            for (Iterator<Entry<K,V>> it : nextIterators) {
+                if (it != null) {
+                    try {
+                        it.remove();
+                    } catch (UnsupportedOperationException uoe) {
+                        e = uoe;
+                    }
+                }
+            }
+            populated = false;
+            if (e != null) {
+                throw new UnsupportedOperationException("One or more of the underlying sets does not support this operation", e);
+            }
+        }
+
+        /* Some utility methods */
+        private boolean equals(K o1, K o2) {
+            if (o1 == null) {
+                return o2 == null;
+            } else if (o2 == null) {
+                return false;
+            } else {
+                if (map.comparator() == null) {
+                    return o1.equals(o2);
+                } else {
+                    return map.comparator().compare(o1, o2) == 0;
+                }
+            }
+        }
+
+        private void populate() {
+            populated = false;
+
+            // update the last value for those iterators contributing to
+            // the last returned value
+            for (int i = 0; i < nextIterators.size(); i++) {
+                if (nextIterators.get(i) != null) {
+                    Iterator<Entry<K,V>> it = nextIterators.get(i);
+                    if (it.hasNext()) {
+                        Entry<K,V> val = it.next();
+                        lastList.set(i, val.getKey());
+                        if ((rewriteStrategy == null) || (!map.containsKey(val.getKey()))
+                                        || (rewriteStrategy.rewrite(val.getKey(), map.get(val.getKey()), val.getValue()))) {
+                            map.put(val.getKey(), val.getValue());
+                        }
+                    } else {
+                        lastList.set(i, null);
+                        finished[i] = true;
+                    }
+                }
+            }
+
+            if (!map.isEmpty()) {
+                nextKey = map.firstKey();
+                nextValue = map.remove(nextKey);
+                for (int i = 0; i < iterators.size(); i++) {
+                    if (!finished[i] && equals(nextKey, lastList.get(i))) {
+                        nextIterators.set(i, iterators.get(i));
+                    } else {
+                        // if the iterator is finished, or did not contribute to the value being returned
+                        // then null it out since the value returned is already in the map to compare
+                        // on the next round
+                        nextIterators.set(i, null);
+                    }
+                }
+                populated = true;
+            }
+        }
+    }
```

Comment:
I'm very unclear on how this method is supposed to work. This could easily be a me problem, but I'm struggling to understand what is going on here and why.

Reply:
We had the conversation off line for this one.  I am presumming this is now understood.  However I will try to add comments to make things more clear.